### PR TITLE
support generic resource action invoke

### PR DESCRIPTION
**Pull request by williexu**
---
Fixes https://github.com/Azure/azure-cli/issues/2607
Samples (on vm resource):
`az resource invoke-action --action powerOff --ids <vm_id>`
`az resource invoke-action --action capture --ids <vm_id> --parameters {"vhdPrefix":"myPrefix","destinationContainerName":"container","overwriteVhds":"true"}`
`az resource invoke-action --action capture --ids <vm_id> --parameters @{file}`

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [x] Each command and parameter has a meaningful description.
- [x] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

### CODE REVIEW

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_help.py [hunk 1]
```diff
@@ -8,68 +8,59 @@
...
     type: group
-    short-summary: Manage Azure Managed Applications.
+    short-summary: Manage managed application definitions.
 """
 helps['managedapp create'] = """
     type: command
-    short-summary: Create a managed application.
+    short-summary: Creates a managed application.
     examples:
-        - name: Create a managed application of kind 'ServiceCatalog'. This requires a valid managed application definition ID.
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_help.py]: Please rebase to get rid of unrelated changes

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py [hunk 2]
```diff
@@ -41,6 +40,9 @@
 
 register_cli_argument('resource list', 'name', resource_name_type)
 register_cli_argument('resource move', 'ids', nargs='+')
+register_cli_argument('resource invoke-action', 'action', help='The action that will be invoked')
+register_cli_argument('resource invoke-action', 'parameters', help='Parameter arguments for the action that will be passed along in the post request body')
+register_cli_argument('resource invoke-action', 'post_url', help='The url that will be sent in the POST request (must include api-version query parameter)')
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py]: Per on going mail thread, we might expose a general request url which you can apply any http method

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py [hunk 3]
```diff
@@ -520,6 +520,23 @@ def tag_resource(tags,
     return res.tag(tags)
 
 
+def invoke_resource_action(action=None, parameters=None, post_url=None,
+                           resource_group_name=None, resource_provider_namespace=None,
+                           parent_resource_path=None, resource_type=None, resource_name=None,
+                           resource_id=None, api_version=None):
+    """ Invokes the provided action on an existing resource. Action specified through either action or post-url"""
+    if post_url:
+        if action or resource_group_name or resource_id or resource_name or resource_type:
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: suggest: `any([action, resource_group_name, resource_id, resource_name, resource_type])`

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py [hunk 4]
```diff
@@ -1359,6 +1376,70 @@ def tag(self, tags):
                                                    self.api_version,
                                                    parameters)
 
+    def invoke_action(self, action=None, parameters=None, url=None):
+        """
+        Formats Url if none provided and sends the POST request with the url and parameters.
+        """
+        from msrestazure.azure_operation import AzureOperationPoller
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: //CC: @lmazuel, we are accessing private fields like following. Feel free to comment if we need adjustments to get things maintainable. We are having e2e test coverage, so should be fine

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py [hunk 5]
```diff
@@ -1359,6 +1376,70 @@ def tag(self, tags):
...
+
+        def get_long_running_status(status_link, headers=None):
+            request = client.get(status_link)
+            if headers:
+                request.headers.update(headers)
+            return client.send(request, header_parameters)
+
+        def get_long_running_output(response):
+            from msrestazure.azure_exceptions import CloudError
+            if response.status_code not in [200, 202, 204]:
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: If we will cover non-post action, let us cross check to make sure we include more valid http return codes

**Comment by davidebbo** [comment]
What about actions that take a body, like `slotsswap` on a `Microsoft.Web/sites` resource? Or is that what `--parameters` is for? Not super clear whether that refers to query string or request body.

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py (hunk 2 above)
Comment (yugangw-msft) [reply src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py]: Sorry for changing the mind, but let us get rid of it.
We can move such functionalities into ms-rest-azure moving forward

**Comment by williexu** [comment]
> What about actions that take a body, like slotsswap on a Microsoft.Web/sites resource? Or is that what --parameters is for? Not super clear whether that refers to query string or request body.

@davidebbo , Yes, that is what `--parameters` is for. It allows the user to specify any JSON to be sent along in the request body. See the following example:
 `az resource invoke-action --action capture --ids <vm_id> --parameters {"vhdPrefix":"myPrefix","destinationContainerName":"container","overwriteVhds":"true"}`. 
Sorry if this wasn't more clear.

**Comment by yugangw-msft** [comment]
@williexu, @davidebbo 
Suggest to rename `--parameters` to `--request-body`? We don't need to name the exactly same like PS. `--parameters` is a too general to me

**Comment by williexu** [comment]
Addressed comments. I have removed the `--post-url` parameter as per discussion with @yugangw-msft . This was included as a way to allow users with niche requests to specify their own url to post. This will instead be considered for ms-rest-azure if needed. 

@davidebbo I changed the name of `--parameters` to `--request-body` instead. Hopefullt this will make the argument more clear.

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/tests/test_resource.py [hunk 6]
```diff
@@ -749,5 +749,28 @@ def test_group_deployment_crossrg(self, resource_group, resource_group_cross):
...
+        username = 'ubuntu'
+        password = self.create_random_name('Longpassword#1', 30)
+
+        vm_json = self.cmd('vm create -g {} -n {} --use-unmanaged-disk --image UbuntuLTS --admin-username {} '
+                           '--admin-password {} --authentication-type {}'
+                           .format(resource_group, vm_name, username, password, 'password')).get_output_in_json()
+
+        vm_id = vm_json.get('id', None)
+
+        self.cmd('resource invoke-action --action powerOff --ids {}'.format(vm_id))
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/tests/test_resource.py]: can you add a `resource show` to verify the machine was indeed powered off?

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py [hunk 7]
```diff
@@ -1266,7 +1277,7 @@ def __init__(self,
 
         self.resource_group_name = resource_group_name
         self.resource_provider_namespace = resource_provider_namespace
-        self.parent_resource_path = parent_resource_path
+        self.parent_resource_path = parent_resource_path if parent_resource_path else ''
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: if you like to do this, then you can remove all occurrences of `self.parent_resource_path or ''` in the same class
Or you back the change out and apply ` self.parent_resource_path or ''` inside `invoke_action`. Either way, let us ensure the code is consistent.

**Comment by davidebbo** [comment]
Sounds, good!

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py [hunk 8]
```diff
@@ -1359,6 +1370,69 @@ def tag(self, tags):
                                                    self.api_version,
                                                    parameters)
 
+    def invoke_action(self, action, request_body):
+        """
+        Formats Url if none provided and sends the POST request with the url and parameters.
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: rename `parameters` to `request_body`

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py [hunk 9]
```diff
@@ -41,6 +40,8 @@
 
 register_cli_argument('resource list', 'name', resource_name_type)
 register_cli_argument('resource move', 'ids', nargs='+')
+register_cli_argument('resource invoke-action', 'action', help='The action that will be invoked')
```
Comment (troydai) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py]: What action? Http action? Should we elaborate in the help?

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py [hunk 10]
```diff
@@ -1359,6 +1370,69 @@ def tag(self, tags):
...
+        """
+        from msrestazure.azure_operation import AzureOperationPoller
+        query_parameters = {}
+        serialize = self.rcf.resources._serialize  # pylint: disable=protected-access
+        deserialize = self.rcf.resources._deserialize  # pylint: disable=protected-access
+        client = self.rcf.resources._client  # pylint: disable=protected-access
+
+        url = '/subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/' \
+            '{resourceProviderNamespace}/{parentResourcePath}/{resourceType}/{resourceName}/{action}'
+        path_format_arguments = {
```
Comment (troydai) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: No need to pass keyword parameter in this form.

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py [hunk 11]
```diff
@@ -1359,6 +1370,69 @@ def tag(self, tags):
...
+        url = client.format_url(url, **path_format_arguments)
+
+        # Construct parameters
+        query_parameters['api-version'] = serialize.query("api_version", self.api_version, 'str')
+
+        # Construct headers
+        header_parameters = {}
+        header_parameters['Content-Type'] = 'application/json; charset=utf-8'
+        if self.rcf.resources.config.generate_client_request_id:
+            header_parameters['x-ms-client-request-id'] = str(uuid.uuid1())
```
Comment (troydai) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: Using `uuid4`

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py (hunk 5 above)
Comment (troydai) [reply src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: Shouldn't all status code between 200 and 299 acceptable?

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py (hunk 9 above)
Comment (tjprescott) [reply src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py]: How would someone know what actions are possible on a given resource?

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_help.py [hunk 12]
```diff
@@ -326,7 +326,7 @@
 
 helps['resource invoke-action'] = """
     type: command
-    short-summary: Invoke an action on the resource. Must provide either the --action parameter or the --post-url parameter
+    short-summary: Invoke an action on the resource. Must provide the --action parameter
```
Comment (tjprescott) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_help.py]: > Must provide the --action parameter 

In that case the parameter should be [Required] and you don't need to state that here.

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py [hunk 13]
```diff
@@ -40,7 +40,7 @@
 
 register_cli_argument('resource list', 'name', resource_name_type)
 register_cli_argument('resource move', 'ids', nargs='+')
-register_cli_argument('resource invoke-action', 'action', help='The action that will be invoked')
+register_cli_argument('resource invoke-action', 'action', help='The action that will be invoked on the specified resource')
```
Comment (tjprescott) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py]: That doesn't really make it much more clear. How will someone know what the possible actions are for a resource?  Is there another endpoint that documents this?

Comment (yugangw-msft) [reply src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/_params.py]: This is a generic command as a stop gap where there is no strong type CLI command for that. So it targets to advance users or who are suggested by such people. Under that context, people are expected to check out the rest api themselves which are well documented in msdn, like
https://docs.microsoft.com/en-us/rest/api/compute/virtualmachines/virtualmachines-start.
We should call out such sources in the help though.

**Comment by leidegre** [comment]
Also `--request-body` should be able to be specified as a file, a command line on Windows can at most be around 8000 characters, that's not that much if you have a big resource with lots of things in it.

**Comment by williexu** [comment]
@leidegre, yes it is possible and probably preferred to specify a filepath for the request body. Use @{file} to utilize this option

**Comment by leidegre** [comment]
@williexu and if I just want to PUT a resource, can I do that to? PUT this resource from this file into Azure? Like `az resource update` but with a file as source.

**Comment by tjprescott** [comment]
@leidegre I believe the command you are looking for is `az resource create`.

**Comment by leidegre** [comment]
@tjprescott I believe you just made my day.

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py (hunk 5 above)
Comment (lmazuel) [reply src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: According to ARM this is enough:
https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#call-action-post

File: src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py [hunk 14]
```diff
@@ -1353,12 +1364,75 @@ def tag(self, tags):
...
+                request.headers.update(headers)
+            return client.send(request, header_parameters)
+
+        def get_long_running_output(response):
+            from msrestazure.azure_exceptions import CloudError
+            if response.status_code not in [200, 202, 204]:
+                exp = CloudError(response)
+                exp.request_id = response.headers.get('x-ms-request-id')
+                raise exp
+            return deserialize('str', response)
```
Comment (lmazuel) [inline src/command_modules/azure-cli-resource/azure/cli/command_modules/resource/custom.py]: `return response.text` would be faster
